README.md
# Cimpress Tagliatelle client

This project contains a thin client library for Cimpress' Tagliatelle service.

## Usage

In order to use the client

`npm install cimpress-tagliatelle --save`

and once the package is available

### Import

`const Tagliatelle = require('cimpress-tagliatelle');`

or 

`import Tagliatelle from 'cimpress-tagliatelle'`

### Instantiation
```
const options = {
   baseUrl: 'string', // Base url of the service, by default it will be https://tagliatelle.trdlnk.cimpress.io
   timeout: '3000', // Default: 3000 ms
   retries: 3, // By default there are 3 retries
   retryDelayInMs: 1000 // Default: 1000
}

const tagliatelle = new Tagliatelle(options);

```
### Class functions 


```javascript
    const accessToken = "ACCESS_TOKEN"
    const resourceId = "https://stereotype.trdlnk.cimpress.io/v1/templates/demo"
    const namespace= "test-namespace"
    const tagId = "urn:namespace:specification"
    const tagValue = "tag_value"

    // Get all the tags asociated to a resource, filtered by namespace if provided.

    tagliatelle
        .getTagsByResource(accessToken, resourceUri, namespace) // The 3rd parameter namespace is optional
        .then(data => {
            console.log(data)
            /*
            {
                count: 2,
                total: 2,
                results: [
                {
                    id: '00000000-0000-0000-0000-000000000000',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                        }
                    }
                },
                {
                    id: '00000000-0000-0000-0000-000000000001',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000001"
                        }
                    }
                },
                ]
            }

            */
        })
    
    // Get all the tags asociated to the resources provided, filtered by namespace if provided.
    const resourcesUris = [
        "https://stereotype.trdlnk.cimpress.io/v1/templates/demo",
        "https://stereotype.trdlnk.cimpress.io/v1/templates/demo2"
    ]
    tagliatelle
        .getTagsByResources(accessToken, resourcesUris, namespace) // The 3rd parameter namespace is optional
        .then(data => {
            console.log(data)
            /*
            {
                count: 2,
                total: 2,
                results: [
                {
                    id: '00000000-0000-0000-0000-000000000000',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                        }
                    }
                },
                {
                    id: '00000000-0000-0000-0000-000000000001',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo2',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000001"
                        }
                    }
                },
                ]
            }

            */
        })
        
    // Get the tag asociated with the key and the resourceUri asociated
    const key = "urn:stereotype:isTest";
    tagliatelle
        .getTagByResourceAndKey(accessToken, resourceUri, key)
        .then(data => {
            console.log(data)
            /*
            {
                id: '00000000-0000-0000-0000-000000000000',
                key: 'urn:stereotype:isTest',
                value: 'true',
                resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                createdAt: '2018-11-01T13:08:11.884Z',
                createdBy: '[SUB]',
                _links: {
                    "self": {
                        "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                    }
                }
            }
            */
        })
    
    
    // Get all the tags with the key provided
    tagliatelle
        .getTagsByKey(accessToken, key)
        .then(data => {
            console.log(data)
            /*
            {
                count: 2,
                total: 2,
                results: [
                {
                    id: '00000000-0000-0000-0000-000000000000',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                        }
                    }
                },
                {
                    id: '00000000-0000-0000-0000-000000000001',
                    key: 'urn:stereotype:isTest',
                    value: 'true',
                    resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo2',
                    createdAt: '2018-11-01T13:08:11.884Z',
                    createdBy: '[SUB]',
                    _links: {
                        "self": {
                            "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000001"
                        }
                    }
                },
                ]
            }

            */
        })
        
    // Get the tag with the id provided
    const id = '00000000-0000-0000-0000-000000000000';
    tagliatelle
        .getTagById(accessToken,id)
        .then(data => {
            console.log(data)
            /*
            {
                id: '00000000-0000-0000-0000-000000000000',
                key: 'urn:stereotype:isTest',
                value: 'true',
                resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                createdAt: '2018-11-01T13:08:11.884Z',
                createdBy: '[SUB]',
                _links: {
                    "self": {
                        "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                    }
                }
            }
            */
        })
    

    // Creates a tag in a specific resource

    const tagValue = true;
    tagliatelle
        .createTag(accessToken, resourceUri, tagKey, tagValue) 
        .then(tag => {
            console.log(tag);
            /*
            {
                id: '00000000-0000-0000-0000-000000000000',
                key: 'urn:stereotype:isTest',
                value: 'true',
                resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                createdAt: '2018-11-01T13:08:11.884Z',
                createdBy: '[SUB]',
                _links: {
                    "self": {
                        "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                    }
                }
            }
            */
        })
            

    // Updates a tag

    const tagValue = true;
    tagliatelle
        .updateTag(accessToken, id, resourceUri, key, tagValue) 
        .then(tag => {
            console.log(tag);
            /*
            {
                id: '00000000-0000-0000-0000-000000000000',
                key: 'urn:stereotype:isTest',
                value: 'true',
                resourceUri: 'https://stereotype.trdlnk.cimpress.io/v1/templates/demo',
                createdAt: '2018-11-01T13:08:11.884Z',
                createdBy: '[SUB]',
                _links: {
                    "self": {
                        "href": "https://tagliatelle.trdlnk.cimpress.io/v0/tags/00000000-0000-0000-0000-000000000000"
                    }
                }
            }
            */
        })

    // Deletes a tag

    const tagValue = true;
    tagliatelle
        .deleteTag(accessToken, id) 
        .then(tag => { /* no-content */})

```

## Support

For any inquiries, we invite you to reach out to the Trdelnik Squad at TrdelnikSquad@cimpress.com.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.5] - 2019.09.10
## [0.3.4] - 2019.07.17
## [0.3.3] - 2019.07.17
### Changed
- Upgrading dependencies to fix vulnerability

## [0.3.2] - 2019.06.23
### Changed
- Bumping axios to fix a vulnerability

## [0.3.1] - 2019.01.17
### Changed
- Timeouts when calling the underlying service should be retried. 

## [0.3.0] - 2018.12.18
### Added
- Added support for debugFunction and errorFunction options. Defaults to console.
- Added support for skipCache option to avoid caching issues. Defaults to true. 

## [0.2.1] - 2018.12.18
### Changed
- Avoid caching issues for GET call

## [0.2.0] - 2018.12.12
### Added
- Add support for offset parameter
- Add support for retrieving paged information

## [0.1.1] - 2018.11.16
### Update
- Add missing dependency babel-runtime

## [0.1.0] - 2018.10.31
### Update
- Added createOrUpdateTag method. 

## [0.0.1] - 2018.10.31
### Update
- Initial version 


__tests__/Tagliatelle_tests.js
'use strict';

const mock = require('xhr-mock').default;
const Tagliatelle = require('../src/Tagliatelle');
const TagNotFound = require('../src/errors/TagNotFound');

describe('Tagliatelle', () => {
    beforeEach(() => mock.setup());
    afterEach(() => mock.teardown());

    describe('Constructor ', () => {
        test('Constructs correctly without params', () => {
            const ts = new Tagliatelle();
            expect(ts.baseUrl).toBe('https://tagliatelle.trdlnk.cimpress.io');
            expect(ts.timeout).toBe(3000);
            expect(ts.retryAttempts).toBe(3);
            expect(ts.retryDelayInMs).toBe(1000);
        });

        test('Constructs correctly with params', () => {
            const params = {
                baseUrl: 'https://tagliatelle2.cimpress.io',
                timeout: 1000,
                retryAttempts: 6,
                retryDelayInMs: 0,
            };
            const ts = new Tagliatelle(params);
            expect(ts.baseUrl).toBe(params.baseUrl);
            expect(ts.timeout).toBe(params.timeout);
            expect(ts.retryAttempts).toBe(params.retryAttempts);
            expect(ts.retryDelayInMs).toBe(params.retryDelayInMs);
        });

        test('Unrecognised param shows console.error', () => {
            console.error = jest.fn();
            const params = {
                fakeParam: 1234,
            };
            const ts = new Tagliatelle(params);
            expect(console.error).toBeCalled();
        });
    });

    describe('getTagByResource(accessToken, resourceUri, namespace)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template';
        const res = [
            {
                'resourceUri': resourceUri,
                'key': 'urn:stereotype:templateType',
                'value': 'xml',
            },
            {
                'resourceUri': resourceUri,
                'key': 'urn:pollAndMail:templateIsActive',
                'value': 'true',
            },
        ];

        test('No namespace specified returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&resourceUri=${encodeURIComponent(resourceUri)}`), {
                status: 200,
                body: res,
            });

            ts.getTagsByResource('MY_TOKEN', resourceUri)
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });

        test('Namespace specified returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&namespace=my-namespace&resourceUri=${encodeURIComponent(resourceUri)}`), {
                status: 200,
                body: res,
            });

            ts.getTagsByResource('MY_TOKEN', resourceUri, 'my-namespace')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });
    });

    describe('getTagsByResources(accessToken, resourcesUriArray, namespace)', () => {
        const resourcesUris = [
            'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template',
            'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template2',
        ];
        const res = [
            {
                'resourceUri': resourcesUris[0],
                'key': 'urn:stereotype:templateType',
                'value': 'xml',
            },
            {
                'resourceUri': resourcesUris[1],
                'key': 'urn:pollAndMail:templateIsActive',
                'value': 'true',
            },
        ];

        test('No namespace specified returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&resourceUri=${encodeURIComponent(resourcesUris[0])}&resourceUri=${encodeURIComponent(resourcesUris[1])}`), {
                status: 200,
                body: res,
            });
            ts.getTagsByResources('MY_TOKEN', resourcesUris)
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });

        test('No namespace specified returns correct values with offset', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&resourceUri=${encodeURIComponent(resourcesUris[0])}&resourceUri=${encodeURIComponent(resourcesUris[1])}&offset=1`), {
                status: 200,
                body: res,
            });
            ts.getTagsByResources('MY_TOKEN', resourcesUris, null, 1)
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });

        test('Namespace specified returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&namespace=my-namespace&resourceUri=${encodeURIComponent(resourcesUris[0])}&resourceUri=${encodeURIComponent(resourcesUris[1])}`), {
                status: 200,
                body: res,
            });

            ts.getTagsByResource('MY_TOKEN', resourcesUris, 'my-namespace')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });
    });


    describe('getTagByResourceAndKey(accessToken, resourceUri, tagKey)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template';
        const key = 'urn:stereotype:templateType';
        const res = {
            'resourceUri': resourceUri,
            'key': key,
            'value': 'xml',
        };

        test('Returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=${encodeURIComponent(key)}&resourceUri=${encodeURIComponent(resourceUri)}`), {
                status: 200,
                body: {
                    results: [res],
                },
            });
            ts.getTagByResourceAndKey('MY_TOKEN', resourceUri, key)
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });

        test('Throws TagNotFound when tag doesn\'t exist', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=${encodeURIComponent(key)}&resourceUri=${encodeURIComponent(resourceUri)}`), {
                status: 200,
                body: {
                    results: [],
                },
            });

            ts.getTagByResourceAndKey('MY_TOKEN', resourceUri, key)
                .catch((err) => {
                    expect(err).toBeInstanceOf(TagNotFound);
                    done();
                });
        });
    });


    describe('getTagsByKey(accessToken, tagKey)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template';
        const tagKey = 'urn:stereotype:templateName';
        const res = {
            'resourceUri': resourceUri,
            'tagKey': tagKey,
            'tagValue': 'true',
        };

        test('Get existing tags', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=${encodeURIComponent(tagKey)}`), {
                status: 200,
                body: {
                    results: [res],
                },
            });

            ts.getTagsByKey('MY_TOKEN', tagKey)
                .then((data) => {
                    expect(data).toEqual({results: [res]});
                    done();
                });
        });

        test('Get existing tags, no results', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=${encodeURIComponent(tagKey)}`), {
                status: 200,
                body: {
                    results: [],
                },
            });

            ts.getTagsByKey('MY_TOKEN', tagKey)
                .then((data) => {
                    expect(data).toEqual({results: []});
                    done();
                });
        });
    });

    describe('getTagById(accessToken, id)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test-template';
        const id = '000000000-0000-0000-0000-000000000000';
        const res = {
            'resourceUri': resourceUri,
            'key': 'urn:stereotype:templateType',
            'value': 'xml',
        };

        test('Returns correct values', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags/${encodeURIComponent(id)}(.*)`), {
                status: 200,
                body: res,
            });
            ts.getTagById('MY_TOKEN', id)
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                })
                .catch(done);
        });

        test('Throws TagNotFound when tag doesn\'t exist', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags/${encodeURIComponent(id)}(.*)`), {
                status: 404,
                body: {
                    data: {results: []},
                },
            });

            ts.getTagById('MY_TOKEN', id)
                .catch((err) => {
                    expect(err).toBeInstanceOf(TagNotFound);
                    done();
                });
        });
    });


    describe('createTag(accessToken, resourceUri, tagId)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test';
        const tagId = 'urn:stereotype:templateName';
        const tagValue = 'test';
        const res = {
            id: 'd3d7e128-c21f-4e57-9838-9ac01c81cd04',
            key: tagId,
            value: tagValue,
            resourceUri: resourceUri,
            createdAt: '2018-11-01T11:30:25.293Z',
            createdBy: 'waad|B13t3vPMnXqfQY8Eu2VlPbu7bR-K5Y7aufebBlqye0E',
            _links: {
                self: {
                    href: 'https://tagliatelle.trdlnk.cimpress.io/v0/tags/d3d7e128-c21f-4e57-9838-9ac01c81cd04',
                },
            },
        };

        test('Creates a tag', (done) => {
            const ts = new Tagliatelle();
            mock.post(`${ts.baseUrl}/v0/tags`, {
                status: 201,
                body: res,
            });

            ts.createTag('MY_TOKEN', resourceUri, tagId, 'test')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });
    });

    describe('createOrUpdateTag(accessToken, resourceUri, tagId)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test';
        const tagId = 'urn:stereotype:templateName';
        const tagValue = 'test';
        const res = {
            id: 'd3d7e128-c21f-4e57-9838-9ac01c81cd04',
            key: tagId,
            value: tagValue,
            resourceUri: resourceUri,
            createdAt: '2018-11-01T11:30:25.293Z',
            createdBy: 'waad|B13t3vPMnXqfQY8Eu2VlPbu7bR-K5Y7aufebBlqye0E',
            _links: {
                self: {
                    href: 'https://tagliatelle.trdlnk.cimpress.io/v0/tags/d3d7e128-c21f-4e57-9838-9ac01c81cd04',
                },
            },
        };

        test('Creates a tag when does not exist', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=urn%3Astereotype%3AtemplateName&resourceUri=https%3A%2F%2Fstereotype.trdlnk.cimpress.io%2Fv1%2Ftemplates%2Ftest`), {
                status: 201,
                body: {results: []},
            });
            mock.post(`${ts.baseUrl}/v0/tags`, {
                status: 201,
                body: res,
            });

            ts.createOrUpdateTag('MY_TOKEN', resourceUri, tagId, 'test')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });

        test('Updates a tag when it does exist', (done) => {
            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags?(.*)&key=urn%3Astereotype%3AtemplateName&resourceUri=https%3A%2F%2Fstereotype.trdlnk.cimpress.io%2Fv1%2Ftemplates%2Ftest`), {
                status: 201,
                body: {results: [res]},
            });
            mock.put(`${ts.baseUrl}/v0/tags/d3d7e128-c21f-4e57-9838-9ac01c81cd04`, {
                status: 201,
                body: res,
            });

            ts.createOrUpdateTag('MY_TOKEN', resourceUri, tagId, 'test')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });
    });

    describe('updateTag(accessToken, id, resourceUri, tagId, tagValue)', () => {
        const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test';
        const tagId = 'urn:stereotype:templateName';
        const tagValue = 'test-2';
        const res = {
            resourceUri: resourceUri,
            key: tagId,
            value: tagValue,
        };

        test('Updates a tag', (done) => {
            const ts = new Tagliatelle();
            mock.put(`${ts.baseUrl}/v0/tags/xx`, {
                status: 200,
                body: res,
            });

            ts.updateTag('MY_TOKEN', 'xx', resourceUri, tagId, 'test')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });
    });

    describe('deleteTag(accessToken, id)', () => {
        test('Delete a tag', (done) => {
            const ts = new Tagliatelle();
            const res = {};
            mock.delete(`${ts.baseUrl}/v0/tags/xx`, {
                status: 204,
                body: res,
            });

            ts.deleteTag('MY_TOKEN', 'xx')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });
    });

    describe('getTagById(accessToken, id)', () => {
        test('Get a tag', (done) => {
            const resourceUri = 'https://stereotype.trdlnk.cimpress.io/v1/templates/test';
            const tagId = 'urn:stereotype:templateName';
            const tagValue = 'test-2';
            const res = {
                resourceUri: resourceUri,
                key: tagId,
                value: tagValue,
            };

            const ts = new Tagliatelle();
            mock.get(new RegExp(`${ts.baseUrl}/v0/tags/xx(.*)`), {
                status: 204,
                body: res,
            });

            ts.getTagById('MY_TOKEN', 'xx')
                .then((data) => {
                    expect(data).toBe(res);
                    done();
                });
        });
    });
});


src/Tagliatelle.js
import axios from 'axios';
import axiosRetry from 'axios-retry';
import TagNotFound from './errors/TagNotFound';
import ConflictError from './errors/ConflictError';
import qs from 'qs';

const DEFAULT_BASE_URL = 'https://tagliatelle.trdlnk.cimpress.io';

class Tagliatelle {
    constructor(clientOptions) {
        const understoodOptions = ['baseUrl', 'timeout', 'retryAttempts', 'retryDelayInMs',
            'autoFetchPagedResults', 'debugFunction', 'errorFunction', 'skipCache'];
        const options = clientOptions || {};
        const timeout = parseInt(options.timeout, 10);
        const retryAttempts = parseInt(options.retryAttempts, 10);
        const retryDelayInMs = parseInt(options.retryDelayInMs, 10);

        this.baseUrl = options.baseUrl || DEFAULT_BASE_URL;
        this.timeout = timeout >= 0 ? timeout : 3000;
        this.retryAttempts = retryAttempts >= 0 ? retryAttempts : 3;
        this.retryDelayInMs = retryDelayInMs >= 0 ? retryDelayInMs : 1000;
        this.autoFetchPagedResults = options.autoFetchPagedResults || false;
        this.skipCache = (typeof options.skipCache === 'boolean') ? options.skipCache : true;
        this.debugFunction = Object.keys(options).includes('debugFunction') ? options.debugFunction : console.info;
        this.errorFunction = Object.keys(options).includes('errorFunction') ? options.errorFunction : console.error;

        Object
            .keys(options)
            .forEach((passedOption) => {
                if (understoodOptions.indexOf(passedOption) === -1) {
                    // eslint-disable-next-line no-console
                    console.error(`[Tagliatelle] Option '${passedOption}' is not understood and will be ignored.`);
                }
            });
    }

    __debug(...args) {
        if (this.debugFunction) {
            this.debugFunction('[TagliatelleClient]', ...args);
        }
    }

    __error(...args) {
        if (this.errorFunction) {
            this.errorFunction('[TagliatelleClient]', ...args);
        }
    }

    __getAxiosInstance(accessToken) {
        const instance = axios.create({
            baseURL: this.baseUrl,
            timeout: this.timeout,
            headers: {
                Authorization: `Bearer ${accessToken.replace('Bearer ', '')}`,
            },
        });

        if (this.retryAttempts > 0) {
            axiosRetry(instance, {
                retries: this.retryAttempts,
                retryDelay: (retryCount) => {
                    return this.retryDelayInMs;
                },
                shouldResetTimeout: true,
            });
        }

        return instance;
    }

    async getTags(accessToken, searchParams) {
        const results = await this.__getTagsPartial(accessToken, searchParams);
        if (!this.autoFetchPagedResults || results.total === results.count + (results.offset||0)) {
            return results;
        }

        const total = results.total;
        const fullResults = results;
        let retrievedCount = results.count;
        while (retrievedCount < total) {
            searchParams.offset = retrievedCount;
            const partial = await this.__getTagsPartial(accessToken, searchParams);
            retrievedCount += partial.count;
            fullResults.count = retrievedCount;
            fullResults.results = fullResults.results.concat(partial.results);
        }
        return fullResults;
    }

    async __getTagsPartial(accessToken, searchParams) {
        let uris = [];
        if (Array.isArray(searchParams.resourceUri) && searchParams.resourceUri.length > 0) {
            uris = uris.concat(searchParams.resourceUri);
        } else if (searchParams.resourceUri) {
            uris.push(searchParams.resourceUri);
        }

        let keys = [];
        if (Array.isArray(searchParams.key) && searchParams.key.length > 0) {
            keys = keys.concat(searchParams.key);
        } else if (searchParams.key) {
            keys.push(searchParams.key);
        }

        const skipCacheParams = {};
        if (this.skipCache) {
            skipCacheParams.skipCache = Math.random();
        }

        const params = Object.assign({}, skipCacheParams, {
            namespace: searchParams.namespace ? searchParams.namespace : undefined,
            key: keys.length > 0 ? keys : undefined,
            resourceUri: uris.length > 0 ? uris : undefined,
            offset: Number.isInteger(Number(searchParams.offset)) && Number(searchParams.offset)>0 ? searchParams.offset : undefined,
        });

        const axiosInstance = this.__getAxiosInstance(accessToken);

        this.__debug(`Calling GET ${this.baseUrl}/v0/tags ...`, params);
        const response = await axiosInstance.get(`${this.baseUrl}/v0/tags`, {
            params: params,
            paramsSerializer: (params) => qs.stringify(params, {indices: false}),
        });

        return response.data;
    }

    async getTagsByResource(accessToken, resourceUri, namespace = null, offset = null) {
        return this.__getTagsPartial(accessToken, {resourceUri, namespace, offset});
    }

    async getTagsByResources(accessToken, resourcesUriArray, namespace = null, offset = null) {
        return this.__getTagsPartial(accessToken, {resourceUri: resourcesUriArray, namespace, offset});
    }

    async getTagByResourceAndKey(accessToken, resourceUri, tagKey, offset = null) {
        return this.__getTagsPartial(accessToken, {resourceUri, key: tagKey, offset})
            .then((res) => {
                const results = res.results;
                if (results.length === 0) {
                    throw new TagNotFound();
                }
                return results[0];
            })
            .catch((e) => {
                this.__error(e);
                throw e;
            });
    }

    async getTagsByKey(accessToken, tagKey, offset = null) {
        return this.__getTagsPartial(accessToken, {key: tagKey, offset});
    }

    async getTagById(accessToken, id) {
        const axiosInstance = this.__getAxiosInstance(accessToken);
        try {
            this.__debug(`Calling ${this.baseUrl}/v0/tags/${encodeURIComponent(id)} ...`);
            const response = await axiosInstance.get(`${this.baseUrl}/v0/tags/${encodeURIComponent(id)}`);
            return response.data;
        } catch (err) {
            this.__error(err);
            if (err.response && err.response.status === 404) {
                throw new TagNotFound();
            }
            throw err;
        }
    }

    async createOrUpdateTag(accessToken, resourceUri, tagKey, tagValue) {
        return this.__getTagsPartial(accessToken, {resourceUri, key: tagKey})
            .then((res) => {
                const results = res.results;
                this.__debug('createOrUpdateTag for ', resourceUri, tagKey, 'found:', results);
                if (results.length === 0) {
                    return this.createTag(accessToken, resourceUri, tagKey, tagValue);
                }
                if (results.length === 1) {
                    return this.updateTag(accessToken, results[0].id, resourceUri, tagKey, tagValue);
                }
                throw new ConflictError('Multiple tags matching.');
            })
            .catch((e) => {
                this.__error(e);
                throw e;
            });
    }

    async createTag(accessToken, resourceUri, tagKey, tagValue) {
        const axiosInstance = this.__getAxiosInstance(accessToken);
        const params = {
            key: tagKey,
            value: tagValue,
            resourceUri: resourceUri,
        };
        try {
            this.__debug(`Calling POST ${this.baseUrl}/v0/tags`, params);
            const response = await axiosInstance.post(`${this.baseUrl}/v0/tags`, params);
            return response.data;
        } catch (e) {
            this.__error(e);
            throw e;
        }
    }

    async updateTag(accessToken, id, resourceUri, tagKey, tagValue) {
        const axiosInstance = this.__getAxiosInstance(accessToken);
        const params = {
            key: tagKey,
            value: tagValue,
            resourceUri: resourceUri,
        };
        try {
            this.__debug(`Calling PUT ${this.baseUrl}/v0/tags/${encodeURIComponent(id)}`, params);
            const response = await axiosInstance.put(`${this.baseUrl}/v0/tags/${encodeURIComponent(id)}`, params);
            return response.data;
        } catch (e) {
            this.__error(e);
            throw e;
        }
    }

    async deleteTag(accessToken, id) {
        const axiosInstance = this.__getAxiosInstance(accessToken);
        try {
            this.__debug(`Calling DELETE ${this.baseUrl}/v0/tags/${encodeURIComponent(id)}`);
            const response = await axiosInstance.delete(`${this.baseUrl}/v0/tags/${encodeURIComponent(id)}`);
            return response.data;
        } catch (e) {
            this.__error(e);
            throw e;
        }
    }
}

module.exports = Tagliatelle;
